src/librustc_resolve/diagnostics.rs

```diff
@@ -1474,21 +1476,49 @@ crate fn add_missing_lifetime_specifiers_label(
                 Applicability::MaybeIncorrect,
             );
         };
-        let suggest_new = |err: &mut DiagnosticBuilder<'_>, sugg| {
+        let suggest_new = |err: &mut DiagnosticBuilder<'_>, sugg: &str| {
             err.span_label(span, "expected named lifetime parameter");
 
-            if let Some(generics) = missing_named_lifetime_spots.iter().last() {
+            for missing in missing_named_lifetime_spots.iter().rev() {
                 let mut introduce_suggestion = vec![];
-                introduce_suggestion.push(match &generics.params {
-                    [] => (generics.span, "<'lifetime>".to_string()),
-                    [param, ..] => (param.span.shrink_to_lo(), "'lifetime, ".to_string()),
+                let msg;
+                let should_break;
+                introduce_suggestion.push(match missing {
+                    MissingLifetimeSpot::Generics(generics) => {
+                        msg = "consider introducing a named lifetime parameter";
+                        should_break = true;
+                        match &generics.params {
+                            [] => (generics.span, "<'a>".to_string()),
+                            [param, ..] => (param.span.shrink_to_lo(), "'a, ".to_string()),
+                        }
+                    }
+                    MissingLifetimeSpot::HRLT { span, span_type } => {
+                        msg = "consider introducing a higher-ranked lifetime";
```

Comment:
A higher-ranked *lifetime* would be something like `for<'a> 'a<X>` -- that is, the lifetime *itself* would be quantified over something (in the case `for<'a> 'a<X>` it would be a type-generic lifetime).

What you would have here is either a higher-ranked type (`for<'a> fn(...)` or `dyn for<'a> Trait<'a>`) or higher-ranked bound (`T: for<'a> MyTrait<'a>`). In both cases they are parameterized in rank-N, where N > 1, by lifetimes, so I would call these "lifetime-generic type / bound".

Reply:
What should the copy be? `consider introducing a higher-ranked type lifetime`? `consider introducing a higher-ranked type`?

Reply:
In the case of bounds:
```
help: consider making the bound lifetime-generic
   |
LL | struct S2<F: for<'a> Fn(&'a i32, &'a i32) -> &'a i32>(F);
   |              ^^^^^^^    ^^^^^^^  ^^^^^^^     ^^^
   = note: for more on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
```
You'll need to substitute "bound" for "type" in the case of `for<'a> fn(...)`

Reply:
Can you check the new output?

Reply:
Seems good. "higher ranked polymorphism" is a bit jargon-y, but it's in a secondary note, so I think it's fine for now.